packages/knip/src/plugins/unbuild/index.ts

```diff
@@ -0,0 +1,26 @@
+import { timerify } from '../../util/Performance.js';
+import { hasDependency } from '../../util/plugin.js';
+import type { IsPluginEnabledCallback, GenericPluginCallback } from '../../types/plugins.js';
+
+// https://github.com/unjs/unbuild#unbuild
+
+export const NAME = 'unbuild';
+
+/** @public */
+export const ENABLERS = ['unbuild'];
+
+export const isEnabled: IsPluginEnabledCallback = ({ dependencies }) => hasDependency(dependencies, ENABLERS);
+
+export const CONFIG_FILE_PATTERNS = ['build.config.{js,cjs,mjs,ts,mts,cts,json}'];
+
+/** @public */
+export const ENTRY_FILE_PATTERNS = [];
+
+/** @public */
+export const PRODUCTION_ENTRY_FILE_PATTERNS = [];
+
+export const PROJECT_FILE_PATTERNS = [];
+
+const findPluginDependencies: GenericPluginCallback = async () => [];
```

Comment:
What would that like look generally? Load plugin -> get the `entries` -> return list of strings/paths? How should I handle object entries - custom builders?

Reply:
At first sight I think it indeed comes down to extract all strings that represent entry paths. E.g. I see there can be a configuration object, or an array of those. Such objects can have `entries` and those can be strings or objects with `input` being the entry path. Those paths is indeed is what the plugin can return from `findDependencies`.

The PostCSS is a bit similar: https://github.com/webpro/knip/blob/main/packages/knip/src/plugins/postcss/index.ts#L26-L33

I personally like to set the types first, like (hopefully) all plugins do, e.g.:

```
const localConfig: PostCSSConfig | undefined = await load(configFilePath);
```

Then iterate from here. You know, Types Driven Development (TDD) 😁

One more remark: some plugins use the `toEntryPattern` helper. This is the way to indicate that the plugin is sure it's an entry path (and not something else, e.g. a package name). So doing an extra `.map(toEntryPattern)` eventually would be nice.

Reply:
Awesome yeah thanks for the advice. I don't need to do the full config types though right? Just the top level / important stuff I think, lots of it isn't useful and it's a very complex config type otherwise.

Reply:
Yes, the minimum to get the job done would be great. In some rare cases I've added the actual types from more extensive types such as `@types/webpack`, but usually the number of relevant types is rather small so the extra dependency is not necessary.

Reply:
![Screenshot 2024-01-08 at 09 19 06 (Code)](https://github.com/webpro/knip/assets/47499684/743b4458-513e-4f38-87f8-8eb3032609b0)

Is Knip doing something to my types? It looks like it is adding `(() => IterableIterator< [number, UnbuildConfigObject]>)`...

The type is just this though:

```ts
export type UnbuildConfigObject = {
  name?: string;
  entries?:
    | string[]
    | {
        builder: string;
        input: string;
        outDir: string;
      }[];
  outDir?: string;
  declaration?: boolean;
  rollup?: Record<string, unknown>;
};

export type UnbuildConfig = UnbuildConfigObject | UnbuildConfigObject[];
```

Reply:
I think the TS error means you can't `map()` over something that is an array OR `undefined`. Looks like the logic does not take into account yet that `entries` can be `undefined`.

> Is Knip doing something to my types?

No, there's nothing in between here :)

Reply:
Or use `??` instead of `||`

Reply:
![Screenshot 2024-01-08 at 09 33 40 (Code)](https://github.com/webpro/knip/assets/47499684/d5f7647e-c0dd-4670-9f73-90d9c362ba7e)

I'm still getting this error though even after making sure that `localConfig.entries` is not undefined...

Reply:
I'm not sure where this `(() => IterableIterator< [number, UnbuildConfigObject]>)` is coming from!

Reply:
```ts
export type UnbuildConfig = UnbuildConfigObject | UnbuildConfigObject[];
```

`localConfig` can be defined, and be an array (iterable), in which case `entries` is not defined.

Reply:
Ooh I think I know. I'm stupid, sorry. 🤦